fix ttm/thermal pull request

bbbaer opened the pull request:
**Summary**

We have modified the original fix ttm code so that it may be used for calculating thermal transport properties in electrical conductors. The new fix allows the ttm grid to have unique electronic properties at each grid point. A more complete description can be found in the docs or in our publication.


**Author(s)**

Bradly Baer (Vanderbilt) 
D Greg Walker (Vanderbilt)  greg.walker@Vanderbilt.Edu

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

No issues.

**Implementation Notes**

Core functionality remains largely the same as fix/ttm. Changes focus on modifying the diffusion equation to use unique properties at each TTM grid point. No negative effects on other parts of LAMMPS.

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

Publication Link : https://link.springer.com/article/10.1007/s00894-025-06433-5

bbbaer committed:
Fix ttm/thermal - A TTM implementation for modeling thermal transport

bbbaer committed:
Fixed a few typos in the docs

akohlmey committed:
add new files in package to .gitignore

akohlmey committed:
replace new potential file with existing one

akohlmey committed:
whitespace and permissions

akohlmey committed:
programming style changes

akohlmey committed:
add version tag

akohlmey commented on src/EXTRA-FIX/fix_ttm_thermal.h:
```diff
@@ -0,0 +1,87 @@
+  double ***k_e_grid;
+
+
+  virtual void allocate_grid();
```
why are these declared virtual? there is no polymorph class derived from it.

bbbaer replied:
These are declared the same way as in the original fix ttm; I followed the same style.

akohlmey commented on src/EXTRA-FIX/fix_ttm_thermal.cpp:
```diff
@@ -0,0 +1,871 @@
+          if (yleft == -1) yleft = nygrid - 1;
+          if (zleft == -1) zleft = nzgrid - 1;
+
+                  // Initialize flags for vacuum
```
This block needs to be properly indented.

bbbaer replied:
Indentation fixed.

akohlmey commented on src/EXTRA-FIX/fix_ttm_thermal.cpp:
```diff
@@ -0,0 +1,871 @@
+                  int down = 1;
+
+                  // Set flags to 0 if vaccum
+                  if (T_electron[iz][iy][xleft] < 1e-5) left = 0;
```
The 1e-5 constant is a "magic number". It is better to declare it as a static constexpr constant at the beginning of the file with a suitable name (and use 1.0e-5 for consistency)

bbbaer replied:
I have replaced my magic number with references to 0 to make the code more clear (as in fix ttm/mod).

akohlmey commented on src/EXTRA-FIX/fix_ttm_thermal.cpp:
```diff
@@ -0,0 +1,871 @@
+            T_electron[iz][iy][ix]*c_e_grid[iz][iy][ix]*del_vol;
+          transfer_energy +=
+            net_energy_transfer_all[iz][iy][ix]*update->dt;
+          //printf("TRANSFER %d %d %d %g\n",ix,iy,iz,transfer_energy);
```
Please remove commented out debug code.

bbbaer replied:
This was present in the original fix ttm as well.  Comments have been removed from fix ttm/thermal.

akohlmey commented on examples/ttm/in.ttm.thermal:
```diff
@@ -0,0 +1,42 @@
+
+fix             1 all nve
+
+fix             twotemp all ttm/thermal 11111 props.in 10 10 10 set 1800.0
```
the file `prop.in` is missing in the example folder

bbbaer replied:
Missing file has been added to the pr.

akohlmey reviewed (changes requested):
@bbbaer Thanks for your submission. I've made some purely formal changes (whitespace, permissions, required additional changes). I replaced the potential file for iron with one that is already in the distribution (examples are supposed to show how to use a command not a benchmark for getting the best possible result).

Comparing fix_ttm.cpp and fix_ttm_thermal.cpp I cannot help but noticing that there is a lot of redundancy. Have you considered making fix ttm/thermal a derived class of fix ttm?

Please note, that I am not checking / validating the science.

There will have to be other reviews.

akohlmey committed:
fix typo

bbbaer committed:
added missing example file

bbbaer committed:
fixes in response to comments

bbbaer commented:
@akohlmey Thank you for your comments. I think I have resolved all of them. 

We did not consider making it a derived class. If you think this would be a valuable change I can run it by my PI and see what his thoughts are.

sjplimp reviewed (changes requested):
I would like to review this PR when it is (nearly) ready to merge.
The previous Q about a stand-alone versus derived class is a good one.
I will also want to know whether it could derive from fix ttm/grid
instead of (or in addition to) fix ttm, since the grid variant is more scalable.

akohlmey committed:
Merge branch 'develop' into bbbaer/develop

akohlmey committed:
add log files for fix ttm/thermal example

akohlmey committed:
refactor fix ttm, ttm/grid, and ttm/thermal so ttm/thermal is derived from ttm

akohlmey committed:
update logs for fix ttm and fix ttm/grid after changing EAM/fs potential

akohlmey committed:
whitespace

akohlmey commented:
@sjplimp I have refactored the pull request code to be derived from FixTTM and removed (most) replicated code. The logfile output is identical to the standalone version.
I have not tried to implement a fix ttm/thermal/grid that could be done at a later point.

bbbaer commented:
> @sjplimp I have refactored the pull request code to be derived from FixTTM and removed (most) replicated code. The logfile output is identical to the standalone version. I have not tried to implement a fix ttm/thermal/grid that could be done at a later point.

Thank you for refactoring, I appreciate the help. I ran the test scripts I used in development and got the same output for all of them as well. 

I reviewed the ttm/grid source and I think that ttm/thermal should be compatible with that approach but the implementation is beyond my skill. I will mention that our and other's work has used 1D grids of 1x1xN ttm cells, which is why we did not consider the scaling issues.

akohlmey committed:
Merge branch 'develop' into develop

akohlmey committed:
address (minor) issues reported by GitHub Copilot

akohlmey committed:
Merge branch 'develop' into develop

sjplimp commented:
Thanks @bbbaer for contributing this enhancement to TTM.  Thanks @akohlmey for refactoring it to be a derived class.
